File: frontend/src/utils/layout/utils/chainBuilder.js
/**
 * Chain Builder Utilities
 * Responsible for grouping chains into disconnected subgraphs (families)
 * and preparing them for independent layout processing.
 */

/**
 * Decomposes the list of nodes into linear "chains" (sequences of nodes).
 * A chain is formed when nodes have 1:1 predecessor/successor relationships
 * and no visual overlap logic violations.
 * 
 * @param {Array<Object>} nodes - All nodes
 * @param {Array<Object>} links - All links
 * @returns {Array<Object>} List of chain objects
 */
export function buildChains(nodes, links) {
    const nodeMap = new Map(nodes.map(n => [n.id, n]));
    const preds = new Map();
    const succs = new Map();

    // specific check for 1-to-1ness
    links.forEach(l => {
        if (!preds.has(l.target)) preds.set(l.target, []);
        if (!succs.has(l.source)) succs.set(l.source, []);
        preds.get(l.target).push(l.source);
        succs.get(l.source).push(l.target);
    });

    const chains = [];
    const visited = new Set();

    // Link Map for fast lookup of {source, target} -> {year, type}
    const linkMap = new Map();
    links.forEach(l => {
        linkMap.set(`${l.source}-${l.target}`, {
            year: l.year || l.event_year,
            type: l.event_type || l.type
        });
    });

    // Helper: Is this a "Primary Continuation" based on time? (Hand-off)
    // A continuation is primary if:
    // 1. It is a LEGAL_TRANSFER
    // OR 2. The link alignment and node boundaries suggest a direct hand-off.
    const isPrimaryContinuation = (parentId, childId) => {
        const link = linkMap.get(`${parentId}-${childId}`);
        const childNode = nodeMap.get(childId);
        const parentNode = nodeMap.get(parentId);

        if (!link || !childNode || !parentNode) return false;

        // Rule: LEGAL_TRANSFER is always a continuation candidate
        if (link.type === "LEGAL_TRANSFER") return true;

        const linkYear = link.year;
        const childStart = childNode.founding_year;
        const parentEnd = getEndYear(parentNode);

        // If link year is available, it MUST match the child birth
        if (linkYear !== undefined) {
            if (Math.abs(linkYear - childStart) > 1) return false;
            // Furthermore, for it to be a "hand-off" (vs a mid-life split),
            // it should be near the parent's death.
            if (Math.abs(linkYear - parentEnd) > 1 && linkYear < parentEnd) return false;
            return true;
        }

        // Fallback: No link year. Compare node boundaries directly.
        // A direct hand-off is when child starts near parent end.
        return Math.abs(childStart - parentEnd) <= 1;
    };

    const currentYear = new Date().getFullYear();

    const getEndYear = (node) => {
        if (!node) return currentYear;
        if (node.dissolution_year) {
            return node.dissolution_year;
        }
        // Zombie Node Check: If no dissolution but has eras, use the last era
        const eras = node.eras || [];
        if (eras.length > 0) {
            const maxEraYear = Math.max(...eras.map(e => e.year || 0));
            if (maxEraYear > 0) return maxEraYear;
        }
        return currentYear;
    };

    /**
     * Identify the unique 'chosen' successor that continues this node's chain.
     */
    const getChosenSuccessor = (nodeId) => {
        const sIds = succs.get(nodeId) || [];
        if (sIds.length === 0) return null;

        // Filter for candidates that are primary continuations
        const candidates = sIds.filter(sId => isPrimaryContinuation(nodeId, sId));

        if (candidates.length === 0) return null;
        if (candidates.length === 1) return candidates[0];

        // Resolve Ambiguity: Check for unique LEGAL_TRANSFER among candidates
        const legalCandidates = candidates.filter(sId => {
            const link = linkMap.get(`${nodeId}-${sId}`);
            return link && link.type === "LEGAL_TRANSFER";
        });

        if (legalCandidates.length === 1) return legalCandidates[0];
        return null;
    };

    /**
     * Identify the unique 'chosen' predecessor that this node continues the chain from.
     */
    const getChosenPredecessor = (nodeId) => {
        const pIds = preds.get(nodeId) || [];
        if (pIds.length === 0) return null;

        // Filter for candidates that are primary continuations
        const candidates = pIds.filter(pId => isPrimaryContinuation(pId, nodeId));

        if (candidates.length === 0) return null;
        if (candidates.length === 1) return candidates[0];

        // Resolve Ambiguity: Check for unique LEGAL_TRANSFER among candidates
        const legalCandidates = candidates.filter(pId => {
            const link = linkMap.get(`${pId}-${nodeId}`);
            return link && link.type === "LEGAL_TRANSFER";
        });

        if (legalCandidates.length === 1) return legalCandidates[0];
        return null;
    };

    // 1. Identification of "Chain Starter" nodes
    const isChainStart = (nodeId) => {
        // A node starts a chain if it has no chosen predecessor,
        // OR if its chosen predecessor has a different chosen successor (split conflict).
        const pId = getChosenPredecessor(nodeId);
        if (!pId) return true;

        // Check if parent chooses US as the primary continuation
        if (getChosenSuccessor(pId) !== nodeId) return true;

        // ALSO: Check for VISUAL overlap with the chosen parent.
        const parentNode = nodeMap.get(pId);
        const myNode = nodeMap.get(nodeId);

        const parentEnd = getEndYear(parentNode);

        // Visual overlap check: parent renders to (parentEnd + 1), so check if that > myStart
        if (parentEnd + 1 > myNode.founding_year) return true; // Visual overlap, break chain

        return false; // No visual overlap, can continue chain
    };

    nodes.forEach(node => {
        if (visited.has(node.id)) return;

        if (isChainStart(node.id)) {
            const chainNodes = [];
            let curr = node.id;

            while (curr) {
                visited.add(curr);
                chainNodes.push(nodeMap.get(curr));

                // Find the unique logical continuation
                const nextId = getChosenSuccessor(curr);
                if (!nextId) break;

                // Symmetry check: handle merges
                // The next node MUST choose MUST choose US as its primary predecessor
                if (getChosenPredecessor(nextId) !== curr) break;

                // Continue
                // CRITICAL: Check for VISUAL overlap (Dirty Data Protection)
                const currNode = nodeMap.get(curr);
                const nextNode = nodeMap.get(nextId);

                const currEnd = getEndYear(currNode);

                if (currEnd + 1 > nextNode.founding_year) {
                    // Visual overlap detected - break chain
                    break;
                }

                curr = nextId;
            }

            chains.push({
                id: chainNodes[0].id,
                nodes: chainNodes,
                startTime: chainNodes[0].founding_year,
                endTime: getEndYear(chainNodes[chainNodes.length - 1]),
                yIndex: 0 // to be assigned
            });
        }
    });

    // Catch-up: If any nodes remain unvisited
    nodes.forEach(node => {
        if (!visited.has(node.id)) {
            visited.add(node.id);
            chains.push({
                id: node.id,
                nodes: [node],
                startTime: node.founding_year,
                endTime: getEndYear(node),
                yIndex: 0
            });
        }
    });

    return chains;
}

/**
 * Group chains into connected components (families) based on links between nodes.
 * 
 * @param {Array<Object>} chains - Array of chain objects
 * @param {Array<Object>} links - Array of link objects {source: nodeId, target: nodeId}
 * @returns {Array<Object>} list of families { chains: [], minStart: number }
 */
export function buildFamilies(chains, links) {
    // Group chains into connected components
    const chainMap = new Map(); // nodeId -> chainId
    chains.forEach(c => c.nodes.forEach(n => chainMap.set(n.id, c)));

    const adj = new Map();
    chains.forEach(c => adj.set(c, new Set()));

    links.forEach(l => {
        // Logic from LayoutCalculator:
        // Nodes are referenced by ID in links.
        const c1 = chainMap.get(l.source);
        const c2 = chainMap.get(l.target);
        if (c1 && c2 && c1 !== c2) {
            adj.get(c1).add(c2);
            adj.get(c2).add(c1);
        }
    });

    const families = [];
    const visited = new Set();

    chains.forEach(c => {
        if (visited.has(c)) return;
        const familyChains = [];
        const q = [c];
        visited.add(c);

        let minStart = c.startTime;

        while (q.length) {
            const cur = q.shift();
            familyChains.push(cur);
            if (cur.startTime < minStart) minStart = cur.startTime;

            adj.get(cur).forEach(neighbor => {
                if (!visited.has(neighbor)) {
                    visited.add(neighbor);
                    q.push(neighbor);
                }
            });
        }

        families.push({
            chains: familyChains,
            minStart
        });
    });

    // Sort families by start year (Gantt style), then by size (largest first)
    families.sort((a, b) => {
        if (a.minStart !== b.minStart) return a.minStart - b.minStart;
        // Tie-breaker: larger families first (optional preference)
        if (a.chains.length !== b.chains.length) return b.chains.length - a.chains.length;
        // Tie-breaker: ID stability (assuming chains[0] exists)
        return a.chains[0].id.localeCompare(b.chains[0].id);
    });

    return families;
}

File: backend/app/optimizer/chain_builder.py
"""
Chain Builder for Backend Optimizer

Ports the frontend's chainBuilder.js logic to Python.
This ensures the optimizer works with the same concept of "chains"
as the frontend, preventing mismatch when applying cached layouts.

Rules for chain building:
1. Strict 1:1 predecessor/successor relationship
2. No visual overlap (parentEnd + 1 > childStart breaks chain)
3. Merge points (>1 predecessor) start a new chain
"""
from typing import List, Dict, Any, Set, Optional
from datetime import datetime


def build_chains(
    nodes: List[Dict[str, Any]], 
    links: List[Dict[str, Any]],
    current_year: Optional[int] = None
) -> List[Dict[str, Any]]:
    """
    Decompose nodes into linear "chains" (sequences of nodes).
    
    A chain is formed when nodes have 1:1 predecessor/successor relationships
    and no visual overlap.
    
    Args:
        nodes: List of node dicts with id, founding_year, dissolution_year
        links: List of link dicts with parentId, childId
        current_year: Optional current year for nodes without dissolution_year
        
    Returns:
        List of chain dicts with id, nodes, startTime, endTime
    """
    if current_year is None:
        current_year = datetime.now().year
    
    # Build node lookup map
    node_map: Dict[str, Dict] = {n["id"]: n for n in nodes}
    
    # Build predecessor and successor maps
    preds: Dict[str, List[str]] = {n["id"]: [] for n in nodes}
    succs: Dict[str, List[str]] = {n["id"]: [] for n in nodes}
    
    for link in links:
        parent_id = link["parentId"]
        child_id = link["childId"]
        
        if parent_id in node_map and child_id in node_map:
            preds[child_id].append(parent_id)
            succs[parent_id].append(child_id)
    
    chains: List[Dict[str, Any]] = []
    visited: Set[str] = set()

    from app.models.enums import LineageEventType

    # Link lookup map: (parent, child) -> link object
    link_map: Dict[tuple, Dict] = {}
    for link in links:
        link_map[(link["parentId"], link["childId"])] = link
    
    def get_end_year(node: Dict) -> int:
        """
        Get the visual end year of a node (dissolution_year, last era, or current_year).
        Matches frontend logic: if no dissolution, try to use last era.
        """
        if node.get("dissolution_year"):
            return node["dissolution_year"]
            
        # Zombie Node Check: If no dissolution but has eras, use the last era
        eras = node.get("eras") or []
        if eras:
            # Safely find max year to avoid sorting assumption
            try:
                # Handle both dict objects and potentially other formats if data varies
                max_era_year = max(e.get("year", 0) for e in eras)
                if max_era_year > 0:
                    return max_era_year
            except Exception:
                pass
                
        return current_year

    def is_primary_continuation(parent_id: str, child_id: str) -> bool:
        """
        Determine if the link is a 'primary' temporal continuation.
        Used to resolve merges: if one parent connects at the child's birth,
        it claims the chain, while others are treated as secondary mergers.
        """
        link = link_map.get((parent_id, child_id))
        child = node_map.get(child_id)
        
        if not link or not child:
            return False
            
        link_year = link.get("year")
        child_start = child.get("founding_year")
        
        if link_year is None or child_start is None:
            return False
            
        # Tolerance: Link matches child start year (allowing for +/- 1 year fuzziness)
        # e.g. GBC(End 1977) -> Link(1978) -> Malvor(Start 1978)
        return abs(link_year - child_start) <= 1
    
    def get_chosen_successor(node_id: str) -> Optional[str]:
        """
        Identify the unique 'chosen' successor that continues this node's chain.
        - If 1 successor: always return it (Rule 1).
        - If multiple: return the one with LEGAL_TRANSFER if unique (Rule 2).
        """
        s_ids = succs.get(node_id, [])
        if len(s_ids) == 0:
            return None
        if len(s_ids) == 1:
            return s_ids[0]
            
        # Resolve Split: Check for unique LEGAL_TRANSFER
        legal_succs = [s_id for s_id in s_ids 
                       if link_map.get((node_id, s_id), {}).get("type") == LineageEventType.LEGAL_TRANSFER]
        
        if len(legal_succs) == 1:
            return legal_succs[0]
        return None

    def get_chosen_predecessor(node_id: str) -> Optional[str]:
        """
        Identify the unique 'chosen' predecessor that this node continues the chain from.
        - If 1 predecessor: always return it (Rule 1).
        - If multiple: return the one with LEGAL_TRANSFER if unique (Rule 2).
        """
        p_ids = preds.get(node_id, [])
        if len(p_ids) == 0:
            return None
        if len(p_ids) == 1:
            return p_ids[0]
            
        # Resolve Merge: Check for unique LEGAL_TRANSFER
        # We only count those that are also temporally primary continuations
        legal_preds = [p_id for p_id in p_ids 
                       if link_map.get((p_id, node_id), {}).get("type") == LineageEventType.LEGAL_TRANSFER
                       and is_primary_continuation(p_id, node_id)]
        
        if len(legal_preds) == 1:
            return legal_preds[0]
        return None

    def is_chain_start(node_id: str) -> bool:
        """
        Determine if a node should start a new chain.
        """
        # A node starts a chain if it has no chosen predecessor,
        # OR if its chosen predecessor has a different chosen successor (split conflict).
        p_id = get_chosen_predecessor(node_id)
        if not p_id:
            return True
            
        # Check if parent chooses US as the primary continuation
        if get_chosen_successor(p_id) != node_id:
            return True
            
        # Check for visual overlap
        parent_node = node_map.get(p_id)
        my_node = node_map.get(node_id)
        
        if parent_node and my_node:
            parent_end = get_end_year(parent_node)
            my_start = my_node["founding_year"]
            
            # Visual overlap: parent renders to (parent_end + 1), check if that > my_start
            if parent_end + 1 > my_start:
                return True
        
        return False
    
    # Build chains starting from chain-start nodes
    for node in nodes:
        node_id = node["id"]
        
        if node_id in visited:
            continue
        
        if is_chain_start(node_id):
            chain_nodes: List[Dict] = []
            curr = node_id
            
            while curr:
                # Cycle detection / Re-visit check
                if curr in visited:
                    break

                visited.add(curr)
                chain_nodes.append(node_map[curr])
                
                # Find the unique logical continuation
                next_id = get_chosen_successor(curr)
                if not next_id:
                    break
                
                # Symmetry check: handle merges
                # The next node MUST choose MUST choose US as its primary predecessor
                if get_chosen_predecessor(next_id) != curr:
                    break
                
                # Check for visual overlap before continuing
                curr_node = node_map.get(curr)
                next_node = node_map.get(next_id)
                
                if curr_node and next_node:
                    curr_end = get_end_year(curr_node)
                    next_start = next_node["founding_year"]
                    
                    # Visual overlap -> break chain
                    if curr_end + 1 > next_start:
                        break
                
                curr = next_id
            
            if chain_nodes:
                # Use first node's ID as the chain ID (for frontend matching)
                first_node = chain_nodes[0]
                last_node = chain_nodes[-1]
                
                chains.append({
                    "id": first_node["id"],
                    "nodes": chain_nodes,
                    "startTime": first_node["founding_year"],
                    "endTime": get_end_year(last_node),
                })
    
    return chains
